### PR TITLE
Add `active_record.postgresql_adapter_decode_dates` config

### PR DESCRIPTION
**Pull request by JoeDupuis**
Add `active_record.postgresql_adapter_decode_dates` config
to toggle automatic decoding of dates column with the PostgresqlAdapter.

PR #51483 is a breaking change and should have been gated behind a config.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/adapters/postgresql/date_test.rb [hunk 1]
```diff
@@ -39,10 +39,4 @@ def test_bc_date_year_zero
     topic = Topic.create!(last_read: date)
     assert_equal date, Topic.find(topic.id).last_read
   end
-
-  def test_date_decoder
-    date = ActiveRecord::Base.connection.select_value("select '2024-01-01'::date")
-    assert_equal Date.new(2024, 01, 01), date
-    assert_equal Date, date.class
-  end
```
Comment (ghiculescu) [inline activerecord/test/cases/adapters/postgresql/date_test.rb:47]: I think you should keep this test, you just need to turn the config on for the duration of the test. You could also write another test that shows what happens when the config is off.

File: guides/source/upgrading_ruby_on_rails.md [hunk 2]
```diff
@@ -81,6 +81,17 @@ Upgrading from Rails 7.1 to Rails 7.2
...
+
+Manual queries with columns of type date will automatically be decoded to `Date` objects.
+To disable this new behavior and return dates as `String`:
+
+```ruby
+# config/application.rb
+config.active_record.postgresql_adapter_decode_dates = false
+```
+
+
```
Comment (ghiculescu) [inline guides/source/upgrading_ruby_on_rails.md:94]: I don't think this is needed, because by default the config is `false`. It's only `true` if you enable it, and the comments in the new framework defaults file will cover that.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_2.rb.tt [hunk 3]
```diff
@@ -66,3 +66,13 @@
 # on `Post`. With this option enabled, it will also look for a `:comments` association.
 #++
 # Rails.application.config.active_record.automatically_invert_plural_associations = true
+
+###
+# Controls whether the PostgresqlAdapter should decode dates automatically with  manual queries.
```
Comment (ghiculescu) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_2.rb.tt:71]:
```suggestion
# Controls whether the PostgresqlAdapter should decode dates automatically with manual queries.
```

File: activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb [hunk 4]
```diff
@@ -649,7 +649,25 @@ def test_does_not_raise_notice_level_warnings
         end
       end
 
+      def test_date_decoder
```
Comment (ghiculescu) [inline activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb:652]: Could you add another test that shows behaviour when the config is off (the default)?

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 5]
```diff
@@ -1159,8 +1168,8 @@ def add_pg_decoders
             "bool" => PG::TextDecoder::Boolean,
             "timestamp" => PG::TextDecoder::TimestampUtc,
             "timestamptz" => PG::TextDecoder::TimestampWithTimeZone,
-            "date" => PG::TextDecoder::Date,
           }
+          coders_by_name["date"] = PG::TextDecoder::Date if decode_dates
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1172]: Could be another PR, but I think it's worth deprecating this *not* being turned on. So in Rails 7.2 we introduce the new setting but it's opt-in, then in the next Rails version you can no longer opt out. What do you think?

File: railties/test/application/configuration_test.rb [hunk 6]
```diff
@@ -2793,6 +2793,40 @@ def index
...
+
+      app "development"
+
+      assert_equal true, ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.decode_dates
+    end
+
+    test "PostgresqlAdapter.decode_dates is false by default for upgraded apps" do
+      remove_from_config '.*config\.load_defaults.*\n'
+      app_file "config/initializers/active_record.rb", <<~RUBY
+        ActiveRecord::Base.establish_connection(adapter: "postgresql", database: ":memory:")
```
[the remark this answers is not in the document]
Comment (JoeDupuis) [reply railties/test/application/configuration_test.rb:2809]: > just seems like it wouldn't work to me.

It sure doesn't :man_facepalming:. I used this [PR](https://github.com/rails/rails/commit/21a6dbd313efa9fb7feac4d95acaf6271112801f) for reference as the only adapter specific config (still in the codebase). It's a left over I forgot to change.

Interestingly, the test was still valid and passing, as we never query the db in that test. We just need the PostgresqlAdapter constant to be load. 

I'll fix this along with the other comments later today. 

Thank you for the review!

File: activerecord/lib/active_record/railtie.rb [hunk 7]
```diff
@@ -217,6 +217,16 @@ class Railtie < Rails::Railtie # :nodoc:
...
+    initializer "active_record.postgresql_adapter_decode_dates" do
+      config.after_initialize do
+        unless config.active_record.postgresql_adapter_decode_dates
+          ActiveSupport.on_load(:active_record_postgresqladapter) do
+            self.decode_dates = false
+          end
+        end
+      end
+    end
+
```
Comment (JoeDupuis) [inline activerecord/lib/active_record/railtie.rb:229]: @ghiculescu Just a heads up: [I've flipped the default value](https://github.com/rails/rails/pull/51763/files#diff-2468c670eb10c24bd2823e42708489a336d6f21c6efc7e3c4a574166fa77bb22R132) on the adapter itself to true.

It makes this initializer a bit more awkward, but it reduces the spam of deprecation warnings when running tests and fixes [a test that was failing](https://buildkite.com/rails/rails/builds/106954#018f5a97-c2a4-40bd-b855-a72d2d80b729) (false negative) due to the deprecation warning.

Comment (ghiculescu) [reply activerecord/lib/active_record/railtie.rb:229]: Hmm, I don't think we should do this. Because then it's a breaking change for folks who just use Active Record without Rails.

Usually we make the default in AR the non-breaking-change variant, and then we use `load_defaults` / `new_framework_defaults` to opt into the new variant.

Comment (ghiculescu) [reply activerecord/lib/active_record/railtie.rb:229]: Maybe just do it without the deprecation.

Comment (JoeDupuis) [reply activerecord/lib/active_record/railtie.rb]: I see! I'll keep that use case in mind next time. 
I've flipped it back and removed the deprecation.

I'll open another PR for the deprecation if I think of something.

**Comment by JoeDupuis** [comment]
Rebased to fix the flakes.

**Comment by JoeDupuis** [comment]
I rebased over your's to fix the flakes.

**Comment by JoeDupuis** [comment]
@byroot I rebased to fix the failing tests. 
Looks like [there is still a flake](https://buildkite.com/rails/rails/builds/107007#018f64d9-8d94-48fe-b63b-3cab9951fa63) even if Github is showing the test suite passed.

I pushed a attempt at a fix for the flake: https://github.com/rails/rails/pull/51783

**Comment by zzak** [comment]
If you're using AR outside of Rails, this now stops decoding dates entirely, right?

**Comment by JoeDupuis** [comment]
@zzak Yup, without Rails, it would default to the old behavior before [the breaking change was introduced](https://github.com/rails/rails/pull/51483).

I am hoping to add a deprecation warning at some point and remove the old behavior entirely. Adding the deprecation warning in this PR turned out more complex than anticipated. 

Any concern?

**Comment by zzak** [comment]
Nope, sorry I am still catching up. Thank you!
